### DISCUSSION TITLE
UploadFile as a CSV file

### DISCUSSION THREAD

**Discussion by enod** [fastapi/fastapi#8156] [Questions] [answered]
**Description**
I'm trying to upload csv file to fastapi server. Is there any clear way to read fastapi File object and write it as a csv file?

When I receive byte stream through the api, it's difficult to save it as a csv file since there could be comma anywhere in the string. 

One solution could be write incoming bytes files to disk and load it back but I'm wondering if there is a direct approach without writing it into a disk

**Comment by euri10** [comment]
there is a whole section in the docs about uploading files: https://fastapi.tiangolo.com/tutorial/request-files/
if you want to avoid writing to the disk you can use the io module from standard library: https://docs.python.org/3/library/io.html

edit : just note the io module is not async

**Comment by dmontagu** [comment, marked as answer]
To add to @euri10 's comment, I think you will probably want to make use of `UploadFile` to receive the csv file (as described in the docs @euri10 linked). This will transparently handle writing the file to disk if it is too large to store in memory, and has both a sync and async API for saving to disk if desired.

Moreover, If you want to load the csv data _without_ saving to disk, I believe it should be possible to pass it directly to `pandas.read_csv` via:

```python
import pandas as pd

@app.post("/uploadcsv/")
def upload_csv(csv_file: UploadFile = File(...)):
    dataframe = pd.read_csv(csv_file.file)
    # do something with dataframe here (?)
    return {"filename": file.filename}
```

(I haven't tested this, but it seems like it should be supported based on the [pandas docs](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#io-read-csv-table). Other csv readers may require other modifications.)

**Comment by tiangolo** [comment]
Great answers @euri10 and @dmontagu ! Thanks a lot for your help here! :cake: 

@enod yep, right what they said :smile: :point_up:

**Comment by VpkPrasanna** [comment]
Thanks a lot @dmontagu  for the help and it solved the issue quickly :+1: :100:

**Comment by danish45007** [comment]
```
@app.post("/uploadfile")
async def check(file: UploadFile = File(...)):
 
    if file.filename.endswith('.csv'):
        df = pd.read_csv(file.file)
        df_head = df.head(5)
        df_tail = df.tail(5)
        # cli.set("userId",str(df))
        head = df_head.to_json(orient='records')
        tail = df_tail.to_json(orient='records')
        head = eval(head)
        tail = eval(tail)
        # create_db_table()
        insert_data(uuid=1,itemname=file.file.read())
        # final return
        return {
            "head":head,
            "tail":tail
        }

```

**Comment by Kludex** [comment]
What if the CSV has `;` or `<tab>` as delimiter?

EDIT:
Lost a lot of time on it. :cry: 
I'm not sure if this is a potential FastAPI/Starlette issue, or it's from the Pandas side, but look:
```python
    if excel.content_type == "text/csv":
        file_ = deepcopy(excel.file)
        for sep in CSV_SEPARATORS:
            file_ = deepcopy(excel.file)
            try:
                data = pd.read_csv(file_, sep=sep)
            except pd.errors.EmptyDataError:
                continue
```
Without the `deepcopy` is not possible to do it. It means that somehow pandas is modifying the object inside, maybe because it uses as an iterator and iterates over it, then it doesn't iterate anymore after it... I tried using a separator regex, but didn't work as well, I have a `ValueError` exception that I cannot solve as well.

**Comment by tiangolo** [comment]
The default file object is a `SpooledTemporaryFile`, which means that it lives by default in memory and then is put on disk if it's too large. I think it might be a good idea to save the incoming file to a temporary file to make sure it's not in memory that can be lost before processing it further.

**Comment by hua7448** [comment]
thanks a lot 

> To add to @euri10 's comment, I think you will probably want to make use of `UploadFile` to receive the csv file (as described in the docs @euri10 linked). This will transparently handle writing the file to disk if it is too large to store in memory, and has both a sync and async API for saving to disk if desired.
> 
> Moreover, If you want to load the csv data _without_ saving to disk, I believe it should be possible to pass it directly to `pandas.read_csv` via:
> 
> ```python
> import pandas as pd
> 
> @app.post("/uploadcsv/")
> def upload_csv(csv_file: UploadFile = File(...)):
>     dataframe = pd.read_csv(csv_file.file)
>     # do something with dataframe here (?)
>     return {"filename": file.filename}
> ```
> 
> (I haven't tested this, but it seems like it should be supported based on the [pandas docs](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#io-read-csv-table). Other csv readers may require other modifications.)

thanks ! very useful！

**Comment by Atem18** [comment]
For me it does not work, I have the following error : AttributeError: 'SpooledTemporaryFile' object has no attribute 'readable'

Any ideas @tiangolo ?

**Comment by Atem18** [comment]
For your information, I have it working but only if I write the file on disk like the following:

```python
@router.post("/import-ohlc")
async def import_ohlc(
    file: UploadFile = File(...),
    current_user: models.UserTable = Depends(fastapi_users.get_current_user),
):
    with open("/tmp/" + file.filename, "wb") as buffer:
        shutil.copyfileobj(file.file, buffer)
    df = pd.read_csv(
        "/tmp/" + file.filename,
        header=None,
        sep=",",
        dtype=object,
        names=["time", "open", "high", "low", "close", "volume", "trades"],
    )
    df["time"] = pd.to_datetime(df["time"], unit="s")
    s_buf = StringIO()
    df.to_csv(s_buf, index=False)
    s_buf.seek(0)
    conn = engine.raw_connection()
    cursor = conn.cursor()
    cursor.execute("TRUNCATE TABLE backtestohlc")
    cursor.copy_expert("COPY backtestohlc FROM STDIN DELIMITER ',' CSV HEADER", s_buf)
    conn.commit()
    return {"result": "Import successful"}
```

Is there any way to do it without writing the csv on disk before reading it with Pandas ?

I tried to implement a custom class as suggested here : https://stackoverflow.com/questions/47160211/why-doesnt-tempfile-spooledtemporaryfile-implement-readable-writable-seekable
But the dataframe is empty with that method.

Thanks a lot in advance.

**Comment by adriangitvitz** [comment]
@Atem18  you could use file.read() like this

files = await file.read()
df = pd.read_excel(files)

**Comment by stevennic** [comment]
I didn't want to write the file to disk just so I can use pandas. I had to read an Excel file. What worked for me instead was to use `BytesIO` to read it directly from the `SpooledTemporaryFile`:

```python
import pandas as pd
from io import BytesIO

def upload_csv(csv_file: UploadFile = File(...)):
    df = pd.read_excel(BytesIO(file.file.read())
```

**Comment by Pworldpeace** [comment]
from io import StringIO
 
pd.read_csv(StringIO(str(data_file.file.read(), 'utf-8')), encoding='utf-8')

**Comment by stevennic** [comment]
> 
> 
> from io import StringIO
> 
> pd.read_csv(StringIO(str(data_file.file.read(), 'utf-8')), encoding='utf-8')

Correct, my problem was actually similar but for pd.read_excel, so I needed BytesIO

**Comment by GitHunter0** [comment]
@stevennic , @hua7448 , please, could you share what would be the `requests.post(...)` counterpart of your examples?

**Comment by VianneyMI** [comment]
I tried @danish45007 approach to upload a csv file and then convert it to json through pandas and it almost work. I have one issue though, it is introducing escapes (\) in front of each ". 
Exampe:
```
[
    {
        \"key1\":\"value1\"    
    }
]
```
Any idea what is going on ?

**Comment by Biaggio74** [comment]
For me this works ok:

```python
@app.post('/v1/upload-csv/')
async def upload_csvfile(csv_file: UploadFile = File(...)):
    # content = await csv_file.read()
    df = pd.read_csv(
        StringIO(str(csv_file.file.read(), 'utf-8')), encoding='utf-8')
    print(df)
    return {"filename": csv_file.filename,
            "content_type": csv_file.content_type}
```

**Comment by batmanscode** [comment]
> For your information, I have it working but only if I write the file on disk like the following:
> 
> ```python
> @router.post("/import-ohlc")
> async def import_ohlc(
>     file: UploadFile = File(...),
>     current_user: models.UserTable = Depends(fastapi_users.get_current_user),
> ):
>     with open("/tmp/" + file.filename, "wb") as buffer:
>         shutil.copyfileobj(file.file, buffer)
>     df = pd.read_csv(
>         "/tmp/" + file.filename,
>         header=None,
>         sep=",",
>         dtype=object,
>         names=["time", "open", "high", "low", "close", "volume", "trades"],
>     )
>     df["time"] = pd.to_datetime(df["time"], unit="s")
>     s_buf = StringIO()
>     df.to_csv(s_buf, index=False)
>     s_buf.seek(0)
>     conn = engine.raw_connection()
>     cursor = conn.cursor()
>     cursor.execute("TRUNCATE TABLE backtestohlc")
>     cursor.copy_expert("COPY backtestohlc FROM STDIN DELIMITER ',' CSV HEADER", s_buf)
>     conn.commit()
>     return {"result": "Import successful"}
> ```
> 
> Is there any way to do it without writing the csv on disk before reading it with Pandas ?
> 
> I tried to implement a custom class as suggested here : https://stackoverflow.com/questions/47160211/why-doesnt-tempfile-spooledtemporaryfile-implement-readable-writable-seekable But the dataframe is empty with that method.
> 
> Thanks a lot in advance.

I was having issues as well and this is the thing that finally worked for me. I was trying not to save to disk but at least it works for me now lol. I tried all the other methods here :(

Strange thing is, that for one function, `csv: UploadFile = File(...)` and using the file like `csv.file` works but not for the other

For anyone interested here's a [notebook with the code](https://colab.research.google.com/drive/1slm_OlGhdsRZGmMb19AApErZve8S_Xal?usp=sharing) and a [test csv](https://github.com/tiangolo/fastapi/files/7428101/test_02-10-2021.csv) to try it out with

**Comment by rasmikr** [comment]
> I didn't want to write the file to disk just so I can use pandas. I had to read an Excel file. What worked for me instead was to use `BytesIO` to read it directly from the `SpooledTemporaryFile`:
> 
> ```python
> import pandas as pd
> from io import BytesIO
> 
> def upload_csv(csv_file: UploadFile = File(...)):
>     df = pd.read_excel(BytesIO(file.file.read())
> ```
Thanks a lot . It helped

**Comment by hua7448** [comment]
This is an automatic reply, confirming that your email was received.I will reply as soon as possible.Thank you.

**Comment by hubshashwat** [comment]
This works for me.

```
@router.post("/upload_csv")
async def upload_csv(separator=',', encoding='utf-8',
                     csv_file: UploadFile = File()):
    contents = csv_file.file.read()
    buffer = BytesIO(contents)
    data = pd.read_csv(buffer, delimiter=separator, encoding=encoding, nrows=10000)
    buffer.close()
    csv_file.file.close()

```

@batmanscode

**Comment by hua7448** [comment]
This is an automatic reply, confirming that your email was received.I will reply as soon as possible.Thank you.

**Comment by batmanscode** [comment]
> This works for me.
> 
> ```
> @router.post("/upload_csv")
> async def upload_csv(separator=',', encoding='utf-8',
>                      csv_file: UploadFile = File()):
>     contents = csv_file.file.read()
>     buffer = BytesIO(contents)
>     data = pd.read_csv(buffer, delimiter=separator, encoding=encoding, nrows=10000)
>     buffer.close()
>     csv_file.file.close()
> ```
> 
> @batmanscode

Thank you. I didn't realize the files had to be closed.

I ended up using [`csv_file.filename`](https://github.com/batmanscode/pydacc/blob/455c38ed026bec9d6c0ac9811629f49efbff9e63/api.py#L236) and just passing that into [`pd.read_csv`](https://github.com/batmanscode/pydacc/blob/455c38ed026bec9d6c0ac9811629f49efbff9e63/pydacc/data_cleaning.py#L45)
